### Lab 2 - Exercise 6.
#### Modified by: Santiago Carvajal Castañeda

In [1]:
import pandas as pd
import numpy as np

#Dataset from biggie
biggie_df = pd.read_csv('biggie_lyrics.csv',usecols=[1],encoding='latin-1',header=None)
biggie_df.columns = ['lyrics']
biggie_df['lyrics'] = biggie_df['lyrics'].str.replace('[^\w\s]','')
biggie_df['lyrics'] = biggie_df['lyrics'].str.lower()
print("Biggie data")
print(biggie_df)

#Dataset from 2biggie
pac_df = pd.read_csv('2pac_lyrics.csv',usecols=[1],encoding='latin-1',header=None)
pac_df.columns = ['lyrics']
pac_df['lyrics'] = pac_df['lyrics'].str.replace('[^\w\s]','')
pac_df['lyrics'] = pac_df['lyrics'].str.lower()

Biggie data
                                               lyrics
0   fuck all you hoes\nget a grip motherfucker\nye...
1   as i grab the glock put it to your headpiece\n...
2   i dont wanna live no mo\nsometimes i hear deat...
3   to all the ladies in the place with style and ...
4   nineteenseventy somethin nigga i dont sweat th...
5   another day in the ghetto \none look outside i...
6   live from bedfordstuyverson the livest one\nre...
7   uh uh uh cmon\nhah sicker than your average\np...
8   uhh uhhh\nbig po ppa\nno info for the dea\nfed...
9   uhh its the ten crack commandments\nwhat uhh u...
10  the commission\nuncle paulie big ditti\ncaesar...
11  relax and take notes while i take tokes of the...
12  good evenin ladies and gentlemen\nhows everybo...
13  who shot ya\nseperate the weak from the obsole...
14  when i die fuck it i wanna go to hell\ncause i...
15  when the lala hits ya lyrics just splits ya\nh...


In [2]:
biggie_lyrics = biggie_df['lyrics'].values
biggie_lyrics = [song.split('\n') for song in biggie_lyrics]
biggie_lyrics = [line for song in biggie_lyrics for line in song]
pac_lyrics = pac_df['lyrics'].values
pac_lyrics = [song.split('\n') for song in pac_lyrics]
pac_lyrics = [line for song in pac_lyrics for line in song]

rap_lines = []

for line in biggie_lyrics:
    if len(line.split()) > 3:
        rap_lines.append(np.array([0,str(line)]))

for line in pac_lyrics:
    if len(line.split()) > 3:
        rap_lines.append(np.array([1,str(line)]))

rap_lines = np.array(rap_lines)


rap_lines = pd.DataFrame(rap_lines)
rap_lines.columns = ['label','line']
rap_lines.head()
rap_lines['label'] = rap_lines['label'].replace(['0','1'],[0,1])

print(rap_lines)

      label                                               line
0         0                                  fuck all you hoes
1         0                            get a grip motherfucker
2         0  yeah this album is dedicated to all the teache...
3         0  id never amount to nothin to all the people th...
4         0  buildings that i was hustlin in front of that ...
5         0  me when i was just tryin to make some money to...
6         0  and all the niggas in the struggle you know wh...
7         0                   uhha its all good baby baybee uh
8         0                                 it was all a dream
9         0                    i used to read word up magazine
10        0          saltnpepa and heavy d up in the limousine
11        0                         hangin pictures on my wall
12        0     every saturday rap attack mr magic marley marl
13        0                 i let my tape rock til my tape pop
14        0     smokin weed and bamboo sippin on privat

In [3]:
import nltk
nltk.download('punkt')

rap_lines['line'] = rap_lines['line'].apply(nltk.word_tokenize)
print(rap_lines.head())

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Santiago\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
   label                                               line
0      0                             [fuck, all, you, hoes]
1      0                       [get, a, grip, motherfucker]
2      0  [yeah, this, album, is, dedicated, to, all, th...
3      0  [id, never, amount, to, nothin, to, all, the, ...
4      0  [buildings, that, i, was, hustlin, in, front, ...


In [4]:
from nltk.stem import PorterStemmer
stemmer = PorterStemmer()
rap_lines['line'] = rap_lines['line'].apply(lambda x: [stemmer.stem(y) for y in x])
rap_lines.head()

label                                               line
0      0                              [fuck, all, you, hoe]
1      0                         [get, a, grip, motherfuck]
2      0  [yeah, thi, album, is, dedic, to, all, the, te...
3      0  [id, never, amount, to, nothin, to, all, the, ...
4      0  [build, that, i, wa, hustlin, in, front, of, t...

In [5]:
# Converts the list of words into space-separated strings
rap_lines['line'] = rap_lines['line'].apply(lambda x: ' '.join(x))


In [6]:
# Convert a collection of text documents to a matrix of token counts
# http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html
# http://scikit-learn.org/stable/modules/feature_extraction.html#text-feature-extraction
# to allow one letter words count_vect = CountVectorizer(token_pattern = r"(?u)\b\w+\b")
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer(token_pattern = r"(?u)\b\w+\b")  
counts = count_vect.fit_transform(rap_lines['line'])
print (counts)


  (0, 1038)	1
  (0, 2519)	1
  (0, 63)	1
  (0, 859)	1
  (1, 1443)	1
  (1, 943)	1
  (1, 19)	1
  (1, 888)	1
  (2, 1371)	1
  (2, 2278)	1
  (2, 2217)	1
  (2, 2192)	1
  (2, 2220)	1
  (2, 2272)	1
  (2, 569)	1
  (2, 1143)	1
  (2, 58)	1
  (2, 2232)	1
  (2, 2511)	1
  (2, 63)	1
  (3, 24)	1
  (3, 1283)	1
  (3, 1609)	1
  (3, 1515)	1
  (3, 80)	1
  :	:
  (1967, 1909)	1
  (1967, 1213)	1
  (1967, 2118)	1
  (1967, 1253)	1
  (1967, 1531)	1
  (1967, 1371)	1
  (1967, 2220)	1
  (1968, 97)	1
  (1968, 149)	1
  (1968, 1848)	1
  (1968, 1527)	1
  (1968, 856)	1
  (1968, 642)	1
  (1968, 124)	1
  (1968, 83)	1
  (1968, 2220)	1
  (1968, 2272)	1
  (1969, 328)	1
  (1969, 2222)	1
  (1969, 1327)	1
  (1969, 2446)	1
  (1969, 1718)	1
  (1969, 2228)	1
  (1969, 642)	1
  (1969, 1143)	1


In [7]:
counts.shape

(1970, 2535)

In [8]:
# https://stackoverflow.com/questions/28064634/random-state-pseudo-random-numberin-scikit-learn
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(counts, rap_lines['label'], test_size=0.1, random_state=69) 

from sklearn.naive_bayes import MultinomialNB

model = MultinomialNB().fit(X_train, y_train)  

import numpy as np
predicted = model.predict(X_test)
print("Accurcay: ")
print(np.mean(predicted == y_test))
print("Confusion matrix: ")
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test, predicted))

Accurcay: 
0.766497461928934
Confusion matrix: 
[[63 32]
 [14 88]]


## Analyzing the results

Using Shahrokhian code, we obtained 76.6% Accuracy while in N Lidell , they obtained 75.13% Accuracy.
This means that using Shahrokhian code is better. We obtained better results, although, the difference wasn't so big.

In [9]:
per = 0
for i in range(0,10):
    X_train, X_test, y_train, y_test = train_test_split(counts, rap_lines['label'], test_size=0.1) 
    model = MultinomialNB().fit(X_train, y_train)  

    import numpy as np
    predicted = model.predict(X_test)
    print(np.mean(predicted == y_test))
    per += np.mean(predicted == y_test)

    from sklearn.metrics import confusion_matrix
    print(confusion_matrix(y_test, predicted))

print ("average perfromance")
print (per/10.0)

0.7309644670050761
[[63 39]
 [14 81]]
0.7868020304568528
[[58 31]
 [11 97]]
0.7969543147208121
[[ 55  24]
 [ 16 102]]
0.7055837563451777
[[49 37]
 [21 90]]
0.7360406091370558
[[50 38]
 [14 95]]
0.7461928934010152
[[ 46  40]
 [ 10 101]]
0.7208121827411168
[[56 35]
 [20 86]]
0.7309644670050761
[[51 38]
 [15 93]]
0.7461928934010152
[[53 32]
 [18 94]]
0.7360406091370558
[[59 30]
 [22 86]]
average perfromance
0.7436548223350254
